pkg/kubelet/kubelet_pods.go

```diff
@@ -1758,9 +1774,11 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 			ExitCode: 137,
 		}
 
-		// If the pod was not deleted, then it's been restarted. Increment restart count.
-		if pod.DeletionTimestamp == nil {
-			status.RestartCount += 1
+		if !isTerminated {
+			// If the pod was not deleted, then it's been restarted. Increment restart count.
+			if pod.DeletionTimestamp == nil {
```

Comment:
Hrm, I don't know that the deletionTimestamp count is relevant?

Reply:
I think isTerminated subsumes pod.DeletionTimestamp until we get into syncTerminating, so this if block may not show up.

Reply:
That logic was there previously. This is guarding against the case that we get into this block where we're generating a terminated ContainerStatusUnknown---if we know the pod is terminated, we shouldn't increment the restart count.

Reply:
It was there previously as an incorrect substitute for isTerminated.  A pod isn't terminated until the pod worker says it is - so you can have deletion timestamp != nil AND still be running.  For !isTerminated to be true, we have to be in syncPod, and if we're in syncPod, we could still be starting containers, so this execution of syncPod could actually restart a container that when syncPod exits we then terminate.  So this check was historically subtly wrong, and now that we have isTerminated we can reflect that.

Reply:
Looks like if we change this we have unit test failures... I guess we should decide which behaviour we want.

Reply:
Made the change suggested in https://github.com/kubernetes/kubernetes/pull/105411#discussion_r720546489 and updated the unit tests, split that into a separate commit so it can be dropped.